### PR TITLE
Parrot POSIX Directory Semantics

### PR DESCRIPTION
**Pull request by trshaffer**
This is a more complete fix for #1104. These commits try to bring several main services more in line with POSIX behavior. To avoid breaking all the other services, there's a marker for services that can correctly do `O_DIRECTORY` (local, chirp, cvmfs, and irods at the moment). I'm still not confident that all the corner cases and errnos are correct, so I'll test more with these services. All the others shouldn't see a difference from current behavior.

### CODE REVIEW

File: parrot/src/pfs_table.cc [hunk 1]
```diff
@@ -654,7 +650,12 @@ pfs_file * pfs_table::open_object( const char *lname, int flags, mode_t mode, in
 			} else {
 				file = pname.service->open(&pname,flags,mode);
 			}
-			free(fd);
+		} else if(!pname.service->can_open_dirs() && (flags&O_DIRECTORY)) {
+			// This is a hack to maintain compatibility with the interpretation
+			// of O_DIRECTORY used in some of the services. If getdents fails
+			// on certain remote services or similar strangeness,
+			// this is a good place to start looking.
+			file = pname.service->getdir(&pname);
```
Comment (batrick) [inline parrot/src/pfs_table.cc]: Now that we have `can_open_dirs()`, we don't need to implement `getdir` for `local`, `cvmfs`, and `chirp`. I think it's safe to remove the those.

File: parrot/src/pfs_service_irods.cc [hunk 2]
```diff
@@ -92,28 +92,30 @@ class pfs_service_irods : public pfs_service {
 
 		struct irods_file *ifile;
 
-		ifile = irods_reli_open(name->hostport,name->rest,flags,mode);
+		if(!(flags&O_DIRECTORY)) {
+			ifile = irods_reli_open(name->hostport,name->rest,flags,mode);
+		}
 		if(ifile) {
 			return new pfs_file_irods(name,ifile);
+		} if((errno == EISDIR) || (flags&O_DIRECTORY)) {
```
Comment (batrick) [inline parrot/src/pfs_service_irods.cc]: `else if`?

File: parrot/src/pfs_service_cvmfs.cc [hunk 3]
```diff
@@ -1165,82 +1169,73 @@ class pfs_service_cvmfs:public pfs_service {
...
-
-		char **buf = NULL;
-		size_t buflen = 0;
-
-		debug(D_CVMFS, "getdir(%s)", name->rest);
-		int rc = compat_cvmfs_listdir(name->rest, &buf, &buflen);
+		int fd = compat_cvmfs_open(name->rest);
 
-		if(rc<0) return 0;
+		if(fd<0) return 0;
```
Comment (batrick) [inline parrot/src/pfs_service_cvmfs.cc]: This needs a `delete dir;` prior to return. (Missing in the prior code too...)

Comment (batrick) [inline parrot/src/pfs_service_cvmfs.cc]: Whoops wrong, part of the patch! Disregard.

File: parrot/src/pfs_service_cvmfs.cc [hunk 4]
```diff
@@ -1165,82 +1169,73 @@ class pfs_service_cvmfs:public pfs_service {
...
-				}
-			}
-			return dir;
-		}
+			int rc = compat_cvmfs_listdir(name->rest, &buf, &buflen);
 
-		/*
-		Otherwise, go to CVMFS for the directory liting.
-		*/
+			if(rc<0) return 0;
```
Comment (batrick) [inline parrot/src/pfs_service_cvmfs.cc]: Put `pfs_dir *dir = new pfs_dir(name);` below this if statement so we don't leak memory when `compar_cvmfs_listdir` fails.

**Comment by dthain** [comment]
After looking at this PR in more detail, I'm more confused than when I started.

The old behavior was this:

In the Linux system call API, a user can call open() on either a file or a directory.  However, the various remote services tend to distinguish clearly between file opens and directory opens, so they have open(), getdir(), and getlongdir().  Linux open() would eventually reach pfs_table.cc::open_object, which would then examine the open flags (and possibly other things) and then decide whether to invoke service->open or service->getdir.

As I read the new code, it seems like it is combining the open() and getdir() operations in each of the underlying services, and then devolving the check for open flags down to each driver.  This seems counterproductive to me, since it is taking logic that was in one place, and then copying it out to multiple drivers, which now have inconsistent methods.

Looking back at the original bug #1104, my sense is that the problem needs to be solved by adjusting the logic in pfs_table::open_object(), not by refactoring the service modules.  But perhaps I haven't fully understood what's going on.

@trshaffer can you clarify exactly what the bug is here, and how you are attempting to fix it?

**Comment by trshaffer** [comment]
The old behavior was to use the `O_DIRECTORY` flag to determine whether to use `open()` or `getdir()`, which caused problems for odd programs that open directories without `O_DIRECTORY`, hence the original issue. For correct behavior, we need to detect if the path we're opening is *not* a directory and fail if `O_DIRECTORY` was passed, and otherwise return a file/directory depending on what the underlying service does.

I was trying to push this logic down to the services so that they can do these checks under whatever semantics they define. If we assume that no remote service can open directories with `open()`, and that they all fail with appropriate errnos, We can probably keep the `O_DIRECTORY` logic in `pfs_table` and fudge the errnos there.

The other issue is that I didn't want to make too many assumptions about every service, and was worried that changing this at `pfs_table` would introduce more  irregular bugs that only show up down the road on particular services. Thus I was marking major services and trying to maintain previous (incorrect) behavior on the other services.


I'll take another shot at this via `pfs_table` and see if I run into any other issues

**Comment by dthain** [comment]
Ok, that's a helpful explanation, I think we are on the right track.

So the problem is that a Unix program can issue open() and then we don't know whether to expect read() or getdents() next, even though only one of those is possible.

Of the underlying services, I believe that only the local service is the one that would actually support that behavior, right?

So, one approach is that Unix open() could first stat() the underlying file, and then decide whether to open as a file or directory, right?

**Comment by batrick** [comment]
> So the problem is that a Unix program can issue open() and then we don't know whether to expect read() or getdents() next, even though only one of those is possible.
>
>Of the underlying services, I believe that only the local service is the one that would actually support that behavior, right?

This affects all services. If Rodney had run `tar` instead on something in /cvmfs or /chirp, he would have had the same failure. If `tar` opens the directory without `O_DIRECTORY`, then `pfs_table` will open a chirp or cvmfs directory as a regular file and fail.

So: we should not rely on/use `O_DIRECTORY` in `pfs_table`. Instead, the service should figure out whether or not it is opening a directory and return the appropriate `pfs_file` or its subclass (`pfs_dir`). If the flags included `O_DIRECTORY` and the file is not a directory, then the service should return a failure.

>So, one approach is that Unix open() could first stat() the underlying file, and then decide whether to open as a file or directory, right?

We already have local fixed. It does an `open` and then `fstat`:

https://github.com/cooperative-computing-lab/cctools/commit/cf6c32992446850e611855794180fa8b91e8db86

**Comment by batrick** [comment]
>Of the underlying services, I believe that only the local service is the one that would actually support that behavior, right?

Sorry I misread your question. Yes, to my knowledge, `local` (unix) is the only service that returns a handle that can be used for `getdents` or `read`.

**Comment by trshaffer** [comment]
Are the there any restrictions on how remote services implement `open()` and `getdir()`? If we require that remote services fail with `EISDIR` when `open()`ing directories, we might be able to have `pfs_table` look at the combination of `O_DIRECTORY`, `O_WRONLY`/`O_RDWR`, and `errno` to figure out the right thing to return. We would just need to make local a special case

**Comment by dthain** [comment]
Tim you are right, that's what @batrick and I were just discussing.. . He will elaborate here, and then you can take another crack at it.

**Comment by batrick** [comment]
@trshaffer, as @dthain said we think having `service::open` return `EISDIR` is the right way to go to minimize changes to services. I guess the real bug here (from Parrot's perspective) was that `local` would happily open directory as an ordinary file.

So I suggest the following changes:

(1) Modify `pfs_table::open_object` to do `pfs_service::getdir` if `O_DIRECTORY ` is set. Otherwise do `pfs_service::open`. If that fails with `EISDIR`, try `getdir`.
(2) The other loose end then is the `local` driver which needs to properly return `EISDIR`. Go ahead and revert cf6c329's changes and fix `open` to return `EISDIR` when the file is a directory.

**Comment by trshaffer** [comment]
This pile of commits seems to be passing tests. This time around, most logic is in `pfs_table` and we tell local to fail if we try to `open()` a directory.

**Comment by dthain** [comment]
This approach looks good to me; I'll leave it to @batrick for detailed review.

File: parrot/src/pfs_table.cc [hunk 5]
```diff
@@ -660,10 +677,26 @@ pfs_file * pfs_table::open_object( const char *lname, int flags, mode_t mode, in
 				file = pfs_cache_open(&pname,flags,mode);
```
Comment (batrick) [inline parrot/src/pfs_table.cc]: I believe we need to check for `EISDIR` here as well as the other `pfs_cache_open`. At this point, I think it's best to have an `open_directory` function that handles the `flags` logic...

**Comment by trshaffer** [comment]
OK, I just put in checks for `pfs_cache_open`

**Comment by batrick** [comment]
Good work @trshaffer!
